bindings/python/xaynet_sdk/participant.py

```diff
@@ -0,0 +1,254 @@
+
+        Args:
+            self: The participant.
+            global_model: The global model.
+
+        Returns:
+            The `TrainingInput` for `train_round`.
+        """
+        raise NotImplementedError()
+
+    # FIXME: make it possible in the participant state machine to skip a task
+    # def participate_in_sum_task(self) -> bool:
+    #     True
+
+    def participate_in_update_task(self) -> bool:
+        """
+        A callback used by the `InternalParticipant` to determine whether the
+        `train_round` method should be called. This callback is only called
+        if the participant is selected as a update participant. If `participate_in_update_task`
+        returns the `False`, `train_round` will not be called by the `InternalParticipant`.
+
+        If the method is not overridden, it returns `True` by default.
+
+        Returns:
+            Whether the `train_round` method should be called when the participant
+            is an update participant.
+        """
+        return True
+
+    def on_new_global_model(self, global_model: Optional[TrainingInput]) -> None:
+        """
+        A callback that is called by the `InternalParticipant` once a new global model is
+        available. If no global model exists (usually in the first round), `global_model` will
+        be `None`. If a global model exists, `global_model` is already the deserialized
+        global model. (See `deserialize_training_input`)
+
+        If the method is not overridden, it does nothing by default.
+
+        Args:
+            self: The participant.
+            global_model: The deserialized global model or `None`.
+        """
+
+    def on_stop(self) -> None:
+        """
+        A callback that is called by the `InternalParticipant` before the `InternalParticipant`
+        thread is stopped.
+
+        This callback can be used, for example, to show performance values ​​that have been
+        collected in the participant over the course of the training rounds.
+
+        If the method is not overridden, it does nothing by default.
+
+        Args:
+            self: The participant.
+        """
+
+
+class InternalParticipant(threading.Thread):
+    def __init__(
+        self,
+        coordinator_url: str,
+        participant,
+        p_args,
+        p_kwargs,
+        state,
+        scalar,
+    ):
+        # xaynet rust participant
+        self._xaynet_participant = xaynet_sdk.Participant(
+            coordinator_url, scalar, state
+        )
+
+        # https://github.com/python/cpython/blob/3.9/Lib/multiprocessing/process.py#L80
+        # stores the Participant class with its args and kwargs
+        # the participant is created in the `run` method to ensure that the participant/ ml
+        # model is initialized on the participant thread otherwise the participant lives on the main
+        # thread which can created issues with some of the ml frameworks.
+        self._participant = participant
```

Comment:
do we want to use the underscore for private members? some python hardcore users may lose their mind. WDYT? but i'm flex

Reply:
I don't know I thought this is best practice 😅

Reply:
i read a bit of conflicting stuff about it, but mostly about people wanting to make private stuff unaccessible from the outside, which is partially a separate topic. regarding naming convention, i guess we can keep it, sure